docs/development/developer-guide.rst

```diff
@@ -131,3 +131,22 @@ simple Python script ``tools/vendor.py``.
 To refresh the dependencies, run the following command::
 
     $ tox -e vendor
+
+----------------
+Type annotations
+----------------
+
+Setuptools makes use of inferred return annotations to reduce verbosity,
+especially for complex return types. Explicit return types can be added for
+functions whose inferred return type contains ``Any``, or that are not checked by
+mypy (ie.: contains no parameter *and* no return type, see
```

Comment:
I have a curiosity question:

In which instances `mypy` can correctly infer return types without degenerating to `Any`?
For example, I tried the following:

```python
def fn(a: str):
    return int(a)

reveal_type(fn)
```

```console
main.py:4: note: Revealed type is "def (a: builtins.str) -> Any"
Success: no issues found in 1 source file
```

To me the return value should be `int`, but `mypy` infers a "degenerate" `Any`.
Is there any circumstance that it can actually get it right without explicit marks?

Reply:
Oh my... I've always (wrongly) assumed that mypy at least infers the return type for functions it does type-check...

Maybe we should instead turn on some `ANN2` rules for public methods so this can be checked automatically instead of having to add it as a guideline.
- https://docs.astral.sh/ruff/rules/missing-return-type-undocumented-public-function/
- https://docs.astral.sh/ruff/rules/missing-return-type-special-method/
- https://docs.astral.sh/ruff/rules/missing-return-type-static-method/
- https://docs.astral.sh/ruff/rules/missing-return-type-class-method/

https://mypy.readthedocs.io/en/stable/config_file.html#confval-disallow_incomplete_defs is also an option.

Reply:
Even w/o enforcement, running `ruff check pkg_resources --fix --unsafe-fixes --select=ANN2 --ignore=ANN202,ANN204` (ignoring tests) is pretty valuable.

Let's merge #4390 and #4391 first.